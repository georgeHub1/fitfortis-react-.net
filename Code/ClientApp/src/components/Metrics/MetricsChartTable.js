import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Table, Icon, Form, Button, Modal } from 'antd';
import moment from 'moment';
import { EditableCell, EditableContext } from './MetricsEditableCell';
import { MetricsChartDataProps } from '../../containers/Metrics/metrics.chart.propTypes';
import { injectIntl } from 'react-intl';

const sortConst = {
  asc: 'asc',
  none: 'none',
  desc: 'desc'
};

class MetricsChartTable extends PureComponent {
  displayName = MetricsChartTable.name;
  static propTypes = {
    chartDATA: PropTypes.arrayOf(MetricsChartDataProps).isRequired,
    columnsName: PropTypes.object.isRequired,
    deleteMetricDataBulk: PropTypes.func.isRequired,
    upsertEntity: PropTypes.func.isRequired,
    intl: PropTypes.shape({ formatMessage: PropTypes.func.isRequired })
  };
  constructor (props) {
    super(props);
    this.state = {
      confirmModal: false,
      editingKey: null,
      deletedKey: null,
      isDeleting: false,
      sorting: {
        dateTime: sortConst.desc
      }
    };
  }
  getIcon = type => {
    const { sorting } = this.state;

    if (sorting[type] === sortConst.none || !sorting[type]) {
      return '';
    }
    return (sorting[type] === sortConst.desc) ? <Icon type="down" /> : <Icon type="up" />;
  }
  setSorting = column => {
    const { sorting } = this.state;

    if (sorting[column] === sortConst.desc) {
      const obj = { [column]: sortConst.asc };

      this.setState({ sorting: obj, editingKey: null });
    } else {
      const obj = { [column]: sortConst.desc };

      this.setState({ sorting: obj, editingKey: null });
    }
  }

  sortMetricData = chartData => {
    const dateTimeMap = {};
    const { sorting } = this.state;
    const allMetricData = chartData.map(x => x.metricData)
      .reduce((acc, curr) => [...acc, ...curr], []);

    allMetricData.forEach(x => {
      if (dateTimeMap[x.dateTime] === null || (dateTimeMap[x.dateTime] === undefined))
        dateTimeMap[x.dateTime] = {
          key: x.dateTime,
          dateTime: x.dateTime
        };

      dateTimeMap[x.dateTime][x.metricId] = x.value;
      dateTimeMap[x.dateTime][`${x.metricId}comment`] = x.comment;
    });

    const combinedMetricData = Object.values(dateTimeMap);

    let field = null;

    let sort = sortConst.asc;

    Object.entries(sorting).forEach(([k, v]) => {
      if (v && v !== sortConst.none) {
        field = k;
        sort = v;
      }
    });

    if (!field)
      return combinedMetricData;

    const sorted = [...combinedMetricData];

    if (sort === sortConst.desc) {
      sorted.sort((a, b) => {
        if ((a[field] === null || a[field] === undefined) && (b[field] === null && b[field] === undefined))
          return 0;
        if ((a[field] === null || a[field] === undefined))
          return 1;
        if ((b[field] === null || b[field] === undefined))
          return -1;
        if (a[field] > b[field]) return -1;
        if (a[field] === b[field]) return 0;
        return 1;
      });
    }
    else {
      sorted.sort((a, b) => {
        if ((a[field] === null || a[field] === undefined) && (b[field] === null && b[field] === undefined))
          return 0;
        if ((a[field] === null || a[field] === undefined))
          return -1;
        if ((b[field] === null || b[field] === undefined))
          return 1;
        if (a[field] > b[field]) return 1;
        if (a[field] === b[field]) return 0;
        return -1;
      });
    }
    return sorted;
  }
  columns = () => {
    const { chartDATA, intl: { formatMessage } } = this.props;
    const columns = chartDATA.map(metric => {
      return {
        title: () => {
          return (
            <span onClick={() => this.setSorting(metric.metricId)}>
              {metric.name}
              {this.getIcon(metric.metricId)}
            </span>
          );
        },
        dataIndex: metric.metricId,
        key: metric.metricId,
        width: 180,
        editable: true,
        render: (text, record) => {
          return (<span className="editable-cell-value-wrap">{text}</span>);
        }
      };
    });
    const columnsComments = chartDATA.map(metric => {
      return {
        title: () => {
          return (
            <span onClick={() => this.setSorting(`${metric.metricId}comment`)}>
              {`${metric.name} ${formatMessage({ id: 'MetricsChart.Comments'})}`}
              {this.getIcon(`${metric.metricId}comment`)}
            </span>
          );
        },
        dataIndex: `${metric.metricId}comment`,
        key: `${metric.metricId}comment`,
        width: 180,
        editable: true,
        render: (text, record) => {
          return (<span className="editable-cell-value-wrap">{text}</span>);
        }
      };
    });

    return [
      {
        title: (filters, sortOrder) => {
          return (
            <span onClick={() => this.setSorting('dateTime')}>
              {this.props.columnsName.date}
              {this.getIcon('dateTime')}
            </span>
          );
        },
        dataIndex: 'dateTime',
        key: 'dateTime',
        width: 210,
        render: (text, record) => {
          return (<span className="editable-cell-value-wrap">{moment(new Date(text)).format('MM/DD/YYYY hh:mm a')}</span>);
        },
        editable: true
      },
      ...columns,
      ...columnsComments,
      {
        title: '',
        dataIndex: 'Operations',
        key: 'Operations',
        render: (text, record) => {
          const isEditing = this.isEditing(record);

          return isEditing ? (
            <div style={{
              alignItems: 'center',
              display: 'flex',
              justifyContent: 'flex-end'
            }}>
              <EditableContext.Consumer>
                {form => (
                  <Button
                    size="small"
                    onClick={e => {
                      e.preventDefault();
                      e.stopPropagation();
                      this.save(form, record.key);
                    }}
                    shape="circle-outline"
                    type="primary"
                    className="action-btn"
                  >
                    <Icon type="check" style={{ width: 14, height: 14 }} />
                  </Button>
                )}
              </EditableContext.Consumer>
              <Button
                size="small"
                shape="circle-outline"
                className="action-btn"
                onClick={e => {
                  e.preventDefault();
                  e.stopPropagation();
                  this.cancel();
                }}
                type="primary"
              >
                <Icon type="close" style={{ width: 14, height: 14 }} />
              </Button>
            </div>
          ) : (
              <div style={{
                alignItems: 'center',
                display: 'flex',
                justifyContent: 'flex-end'
              }}>
                <Button
                  size="small"
                  shape="circle-outline"
                  className="action-btn"
                  onClick={e => {
                    e.preventDefault();
                    e.stopPropagation();
                    this.onDelete(record);
                  }}
                  type="primary"
                >
                  <Icon type="delete" style={{ width: 14, height: 14 }} />
                </Button>
              </div>);
        }
      }
    ];
  };

  isEditing = record => record.key === this.state.editingKey;
  onConfirmDelete = () => {
    this.setState(
      { isDeleting: true, confirmModal: true },
      () => {
        const { deletedKey } = this.state;
        const { chartDATA, deleteMetricDataBulk } = this.props;
        const result = chartDATA.map(x => x.metricData)
          .reduce((acc, curr) => [...acc, ...curr], [])
          .filter(x => x.dateTime === deletedKey)
          .map(x => x.id);

        deleteMetricDataBulk(result).then(this.cancel);
      }
    );
  }
  onDelete = record => {
    this.setState({ deletedKey: record.key, confirmModal: true });
  }
  cancel = () => {
    this.setState(item => ({ ...item, deletedKey: null, confirmModal: false, isDeleting: false, editingKey: null }));
  };
  save = (form, key) => {
    form.validateFields((error, row) => {
      if (error) {
        return;
      }
      const {
        upsertEntity,
        chartDATA
      } = this.props;
      const dateTime = this.state.editingKey;
      const editedMetricData = chartDATA.map(x => x.metricData)
        .reduce((acc, curr) => [...acc, ...curr], [])
        .filter(x => x.dateTime === dateTime);
      const result = [];
      const newDateTime = row.dateTime.toDate();

      newDateTime.setSeconds(0);
      newDateTime.setMilliseconds(0);
      chartDATA.forEach(x => {
        const point = editedMetricData.find(y => y.metricId === x.metricId);

        if (point) {
          result.push({
            id: point.id,
            date: moment(new Date(newDateTime)).utc().format(),
            metricId: x.metricId,
            value: row[x.metricId],
            comment: row[`${x.metricId}comment`]
          });
        }
        else if (row[x.metricId] !== null && row[x.metricId] !== undefined && row[x.metricId] !== '') {
          result.push({
            date: moment(new Date(newDateTime)).utc().format(),
            metricId: x.metricId,
            value: row[x.metricId],
            comment: row[`${x.metricId}comment`]
          });
        }
      });
      if (!result.length) {
        this.cancel();
      }
      upsertEntity(result, this.cancel);
    });
  }
  edit = key => {
    if (this.state.editingKey && this.state.editingKey !== key) {
      this.setState(item => ({ editingKey: null }), () => {
        this.setState({ editingKey: key });
      });
    }
    else {
      this.setState({ editingKey: key });
    }
  }

  render () {
    const { chartDATA, intl: { formatMessage } } = this.props;
    const { isDeleting, confirmModal } = this.state;
    const sortedData = this.sortMetricData(chartDATA);
    const components = {
      body: {
        cell: EditableCell
      }
    };
    const columns = this.columns().map(col => {
      if (!col.editable) {
        return col;
      }
      let inputTypeValue = null;

      if (col.dataIndex === 'dateTime') {
        inputTypeValue = 'date';
      } else if (col.dataIndex.includes('comment')) {
        inputTypeValue = 'text';
      } else {
        inputTypeValue = 'number';
      }
      return {
        ...col,
        onCell: record => ({
          record,
          inputType: inputTypeValue,
          dataIndex: col.dataIndex,
          title: col.title,
          editing: this.isEditing(record)
        })
      };
    });

    return (<div
      className="metricsTimeLineSection"
      style={{
        borderRadius: '0px',
        padding: '0px',
        width: '97%'
      }}
    >
      <div className={`${sortedData.length === 0 ? 'no-data-table' : ''} detailViewTable`}>
        <EditableContext.Provider value={this.props.form}>
          <Table
            onRow={(record, index, event) => {
              return {
                onClick: e => {
                  this.edit(record.key);
                }
              };
            }}
            rowKey={record => record.key}
            {...this.state}
            columns={columns}
            dataSource={sortedData}
            components={components}
            pagination={{
              pageSize: 5,
              onChange: this.cancel
            }}
            elip
            scroll={{ y: 240}}
          />
          <Modal
            title={formatMessage({ id: 'MetricsChartTable.confirmationCaption' })}
            centered
            className="wrapperModal"
            visible={confirmModal}
            confirmLoading={isDeleting}
            okText={formatMessage({ id: 'MetricsChartTable.confirmationButtonDelete' })}
            cancelText={formatMessage({ id: 'MetricsChartTable.confirmationButtonCancel' })}
            onOk={this.onConfirmDelete}
            cancelButtonProps={{
              disabled: isDeleting
            }}
            onCancel={this.cancel}
          ><p>
              {formatMessage({ id: 'MetricsChartTable.confirmationTextDelete' })}
            </p>

          </Modal>

        </EditableContext.Provider>
      </div>
    </div>
    );
  }
}

export default injectIntl(Form.create()(MetricsChartTable));
