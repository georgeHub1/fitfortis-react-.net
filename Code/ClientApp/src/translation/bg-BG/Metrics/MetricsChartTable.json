{
    "confirmationCaption": "Потвърждение",
    "confirmationButtonDelete": "Изтрий",
    "confirmationButtonCancel": "Отмени",
    "confirmationTextDelete": "Наистина ли искаш да изтриеш този рекорд?"
}