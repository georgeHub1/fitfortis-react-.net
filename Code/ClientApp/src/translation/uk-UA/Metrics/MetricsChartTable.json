{
    "confirmationCaption": "Підтвердження",
    "confirmationButtonDelete": "Видалити",
    "confirmationButtonCancel": "Скасувати",
    "confirmationTextDelete": "Ви впевнені, що хочете видалити цей запис?"
}