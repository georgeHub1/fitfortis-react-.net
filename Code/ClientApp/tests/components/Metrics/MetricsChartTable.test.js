import React from 'react';

import renderer from 'react-test-renderer';
import MetricsChartTable from '../../../src/components/Metrics/MetricsChartTable.js';

const Container = MetricsChartTable;

describe('MetricsChartTable', () => {
  it('renders correctly', () => {
    const wrapper = renderer.create(<Container
      chartDATA={[]}
      columnsName={{}}
      createNewEntry={jest.fn()}
      deleteMetricDataBulk={jest.fn()}
      upsertEntity={jest.fn()}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
