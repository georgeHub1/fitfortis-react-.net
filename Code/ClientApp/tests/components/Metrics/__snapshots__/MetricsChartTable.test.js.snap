// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsChartTable renders correctly 1`] = `
<div
  className="metricsTimeLineSection"
  style={
    Object {
      "borderRadius": "0px",
      "padding": "0px",
      "width": "97%",
    }
  }
>
  <div
    className="no-data-table detailViewTable"
  >
    <div
      className="ant-table-wrapper"
    >
      <div
        className="ant-spin-nested-loading"
      >
        <div
          className="ant-spin-container"
        >
          <div
            className="ant-table ant-table-default ant-table-empty ant-table-fixed-header ant-table-scroll-position-left ant-table-layout-fixed"
            style={Object {}}
          >
            <div
              className="ant-table-content"
            >
              <div
                className="ant-table-scroll"
              >
                <div
                  className="ant-table-header"
                  onScroll={[Function]}
                  style={Object {}}
                >
                  <table
                    className=""
                    style={Object {}}
                  >
                    <colgroup>
                      <col
                        style={
                          Object {
                            "minWidth": 210,
                            "width": 210,
                          }
                        }
                      />
                      <col
                        style={
                          Object {
                            "minWidth": undefined,
                            "width": undefined,
                          }
                        }
                      />
                    </colgroup>
                    <thead
                      className="ant-table-thead"
                    >
                      <tr
                        style={
                          Object {
                            "height": null,
                          }
                        }
                      >
                        <th
                          className="ant-table-row-cell-break-word"
                        >
                          <span
                            className="ant-table-header-column"
                          >
                            <div>
                              <span
                                className="ant-table-column-title"
                              >
                                <span
                                  onClick={[Function]}
                                >
                                  <i
                                    aria-label="icon: down"
                                    className="anticon anticon-down"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      className=""
                                      data-icon="down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                      />
                                    </svg>
                                  </i>
                                </span>
                              </span>
                              <span
                                className="ant-table-column-sorter"
                              />
                            </div>
                          </span>
                        </th>
                        <th
                          className="ant-table-row-cell-last"
                        >
                          <span
                            className="ant-table-header-column"
                          >
                            <div>
                              <span
                                className="ant-table-column-title"
                              >
                                
                              </span>
                              <span
                                className="ant-table-column-sorter"
                              />
                            </div>
                          </span>
                        </th>
                      </tr>
                    </thead>
                  </table>
                </div>
                <div
                  className="ant-table-body"
                  onScroll={[Function]}
                  onWheel={[Function]}
                  style={
                    Object {
                      "maxHeight": 240,
                      "overflowY": "scroll",
                    }
                  }
                  tabIndex={-1}
                >
                  <table
                    className=""
                    style={Object {}}
                  >
                    <colgroup>
                      <col
                        style={
                          Object {
                            "minWidth": 210,
                            "width": 210,
                          }
                        }
                      />
                      <col
                        style={
                          Object {
                            "minWidth": undefined,
                            "width": undefined,
                          }
                        }
                      />
                    </colgroup>
                    <tbody
                      className="ant-table-tbody"
                    />
                  </table>
                </div>
                <div
                  className="ant-table-placeholder"
                >
                  <div
                    className="ant-empty ant-empty-normal"
                  >
                    <div
                      className="ant-empty-image"
                    >
                      <svg
                        height="41"
                        viewBox="0 0 64 41"
                        width="64"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <g
                          fill="none"
                          fillRule="evenodd"
                          transform="translate(0 1)"
                        >
                          <ellipse
                            cx="32"
                            cy="33"
                            fill="#F5F5F5"
                            rx="32"
                            ry="7"
                          />
                          <g
                            fillRule="nonzero"
                            stroke="#D9D9D9"
                          >
                            <path
                              d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                            />
                            <path
                              d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                              fill="#FAFAFA"
                            />
                          </g>
                        </g>
                      </svg>
                    </div>
                    <p
                      className="ant-empty-description"
                    >
                      No Data
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
